user/checkout.man
.\" $Id: checkout.man,v 11.9 2025/09/28 23:02:39 tom Exp $
.TH CHECKOUT 1 2025-09-28 "" "User commands"
.ie n .ds CW R
.el   \{
.ie \n(.g .ds CW CR
.el       .ds CW CW
.\}
.de Es
.ne \\$1
.nr mE \\n(.f
.RS 5n
.sp .7
.nf
.nh
.ta 9n 17n 25n 33n 41n 49n
.ft \*(CW
..
.de Eh
.ft \\n(mE
.fi
.hy \\n(HY
.RE
.sp .7
..
.SH NAME
\fBcheckout\fR \-
RCS check-out utility
.SH SYNOPSIS
\fBcheckout\fI [options] [file-specifications]\fR
.SH DESCRIPTION
\fBCheckout\fR is an extension of the RCS utility \fBco\fR.
It uses a file's check-in date to set the file's modification date
after check-out.
.PP
\fBCheckout\fR uses the rcs utility \fBco\fR.
It is normally invoked from the \fBrcsget\fR script,
but may be invoked in a standalone manner.
\fBCheckout\fR differs from \fBco\fR primarily
in its treatment to the delta date: after invoking \fBco\fR,
\fBcheckout\fR
sets the file's modification date to match the delta-date for
the extracted version.
.SS
Set-UID Operation
Like the \fBcheckin\fR utility, \fBcheckout\fR properly
handles set-uid operation for a non-root effective user.
The archived
files are owned by the effective user, and the checked-out files are
owned by the real user.
.PP
Note that since \fBcheckin\fR initializes the access list
in the archive files, the real user must appear on the access list
to lock a file.
Of course, if no lock is made, then the user need
not be on the access list.
.SH OPTIONS
\fBCheckout\fR recognizes all of the "\fBco\fR" options
except for the "\-\fBd\fR" cutoff date.
Instead, the "\-\fBc\fR"
option is used:
.TP 4n
.B \-c\fR \fIcutoff\fR
specifies the \fIrcs\fR cutoff
date, in SCCS-format
.RS
.Es
YY[MM[DD[HH[MM[SS]]]]]
.Eh
.PP
Punctuation in the date is permitted.
For example,
.Es 2
-c 77/2/2 9:22:25
-c 770202092225
.Eh
.PP
represent the same date.
This option is translated to the \fBco\fR
"\-\fBd\fR" option.
.RE
.PP
Use the "\-\fBx\fR" option of \fBcheckout\fR to direct it
to retain the pathname when translating between working and archive
files, e.g, "\fBpath/name\fR" becomes "\fBpath/RCS/name,v\fR".
If the "\-\fBx\fR" option is not set, "\fBpath/name\fR"
becomes "\fB./RCS/name,v\fR", and "\fBpath/RCS/name,v\fR"
becomes "\fB./name\fR".
.SH OPERATIONS
To extract files using \fBcheckout\fR, you must provide one or
more file names.
For each name, checkout parses the leaf-name and
extracts the specified version from the "./RCS" directory.
.SH ENVIRONMENT
\fBCheckout\fR is a C-language program.
It invokes \fBco\fR
via an \fIexeclp\fR call (so that \fBco\fR must be in your
path), and performs pre- and postprocessing of the archive and working
file to determine the delta-date to use for the extracted file.
Environment variables used by checkout include:
.TP
\fBRCS_DIR\fR
if defined, specifies the directory in which \fIrcs
\fR archive files are found.
Normally files are found in "./RCS".
.TP
\fBTZ\fR
is the POSIX time zone, which is overridden internally
so that file modification dates are independent of the local time
zone.
.SH FILES
Checkout uses the following files:
.TP
\fBco\fR
the RCS check-out program
.SH ANTICIPATED CHANGES
This does not handle \fIrcs\fR branches properly.
.PP
\fBCheckout\fR does not handle the "\-\fBj\fR" and "\-\fBp\fR"
options of \fBco\fR.
.SH AUTHORS
Thomas E. Dickey <dickey@invisible-island.net>
.SH SEE ALSO
checkin, rcsput, ded, ci\ (1), co\ (1)


CHANGES
2025-09-28  Thomas E. Dickey  <dickey@invisible-island.net>

	* package/cm_tools-12.0.spec, package/debian/changelog: bump

	* package/debian/copyright: formatting

	* user/baseline.man, user/checkin.man, user/checkout.man,
	  user/checkup.man, user/copy.man, user/link2rcs.man, user/permit.man,
	  user/rcsput.man:
	check-manpage

	* user/baseline.man, user/checkin.man, user/checkout.man,
	  user/checkup.man, user/copy.man, user/link2rcs.man, user/permit.man,
	  user/rcsput.man:
	fixup-th-man

	* user/rcsget.man, user/vcs.man: check-manpage

	* user/rcsget.man, user/vcs.man: fixup-th-man

	* configure: regen

	* aclocal.m4, configure.in: use CF_PROG_INSTALL

2025-07-09  Thomas E. Dickey  <dickey@invisible-island.net>

	* config.guess, config.sub: update to 2025-07-10

2025-06-08  Thomas E. Dickey  <dickey@invisible-island.net>

	* install-sh: fix license

2025-01-17  Thomas E. Dickey  <dickey@invisible-island.net>

	* package/debian/copyright, COPYING, package/cm_tools-12.0.spec,
	  package/debian/changelog:
	bump

	* configure: regen

	* configure.in: use CF_WITH_SYSTYPE early

2025-01-06  Thomas E. Dickey  <dickey@invisible-island.net>

	* src/checkin/src/checkin.c, src/checkout/src/checkout.c,
	  src/checkup/src/checkup.c, src/checkout/src/rcsget.c,
	  src/copy/src/copy.c, src/link2rcs/src/link2rcs.c,
	  src/permit/src/permit.c, src/vcs/src/delete.c, src/vcs/src/unlock.c,
	  src/checkin/src/rcsput.c, src/baseline/src/baseline.c:
	gcc15-warnings

2024-12-21  Thomas E. Dickey  <dickey@invisible-island.net>

	* aclocal.m4: resync with my-autoconf

2024-07-28  Thomas E. Dickey  <dickey@invisible-island.net>

	* config.guess: 2024/07/27

2024-06-03  Thomas E. Dickey  <dickey@invisible-island.net>

	* config.sub: 2024/05/27

2024-04-25  Thomas E. Dickey  <dickey@invisible-island.net>

	* package/debian/copyright, COPYING, package/cm_tools-12.0.spec,
	  package/debian/changelog:
	bump

	* configure: regen

	* aclocal.m4: resync with my-autoconf

2023-12-27  Thomas E. Dickey  <dickey@invisible-island.net>

	* config.sub: 2023-12-27

2023-12-02  Thomas E. Dickey  <dickey@invisible-island.net>

	* config.guess: 2023-12-02

2023-01-22  Thomas E. Dickey  <dickey@invisible-island.net>

	* package/cm_tools-12.0.spec: bump

	* configure: regen

	* bin/makefile.in:
	build-fix for Solaris: rename variable $(USER) to $(AS_USER)

	* package/debian/copyright, COPYING: bump

	* aclocal.m4, configure.in: use pkg-config to find td_lib

	* package/debian/changelog: bump

2023-01-20  Thomas E. Dickey  <dickey@invisible-island.net>

	* package/debian/compat: bump

2022-12-31  Thomas E. Dickey  <dickey@invisible-island.net>

	* package/cm_tools-12.0.spec, package/debian/changelog: bump

	* aclocal.m4: build-fixes for NetBSD

	* configure: regen

2022-10-11  Thomas E. Dickey  <dickey@invisible-island.net>

	* user/makefile.in: use install-sh if needed

	* package/cm_tools-12.0.spec, package/debian/changelog,
	  package/debian/copyright, COPYING:
	bump

	* configure: regen

	* aclocal.m4: resync with my-autoconf

2022-08-01  Thomas E. Dickey  <dickey@invisible-island.net>

	* config.guess, config.sub: 2022-08-01

2021-03-28  Thomas E. Dickey  <dickey@invisible-island.net>

	* package/cm_tools-12.0.spec, package/debian/changelog: bump

	* configure: regen

	* aclocal.m4: resync with my-autoconf

2021-01-10  Thomas E. Dickey  <dickey@invisible-island.net>

	* configure: regen

	* aclocal.m4: quoting fixes for shellcheck

	* configure: regen

	* aclocal.m4: resync with my-autoconf

	* COPYING, package/cm_tools-12.0.spec, package/debian/changelog,
	  package/debian/copyright:
	bump

2020-12-08  Thomas E. Dickey  <dickey@invisible-island.net>

	* package/cm_tools-12.0.spec, package/debian/changelog: bump

	* user/makefile.in, src/vcs/src/makefile.in,
	  src/baseline/test/makefile.in, src/baseline/makefile.in,
	  src/checkin/test/makefile.in, src/checkout/src/makefile.in,
	  src/checkout/test/makefile.in, src/checkup/src/makefile.in,
	  src/checkup/test/makefile.in, src/copy/src/makefile.in,
	  src/copy/test/makefile.in, src/link2rcs/src/makefile.in,
	  src/link2rcs/test/makefile.in, support/makefile.in:
	use VPATH

2020-12-07  Thomas E. Dickey  <dickey@invisible-island.net>

	* src/checkin/makefile.in, src/checkout/makefile.in,
	  src/checkup/makefile.in, src/copy/makefile.in,
	  src/link2rcs/makefile.in, src/permit/makefile.in, src/vcs/makefile.in:
	use VPATH

	* package/debian/compat: bump

	* src/baseline/src/makefile.in, src/checkin/src/makefile.in,
	  src/permit/src/makefile.in, src/permit/test/makefile.in,
	  src/vcs/test/makefile.in:
	use VPATH

2020-12-06  Thomas E. Dickey  <dickey@invisible-island.net>

	* package/debian/copyright, COPYING: bump

	* configure: regen

	* aclocal.m4: resync with my-autoconf

	* configure.in: without-X

	* package/cm_tools-12.0.spec, package/debian/changelog: bump

2020-11-30  Thomas E. Dickey  <dickey@invisible-island.net>

	* install-sh: update to 2020-11-14

2020-08-16  Thomas E. Dickey  <dickey@invisible-island.net>

	* config.guess, config.sub: 2020-08-17

2019-12-21  Thomas E. Dickey  <dickey@invisible-island.net>

	* package/cm_tools-12.0.spec, package/debian/changelog: bump

	* user/checkup.man: fix typo

2019-12-06  Thomas E. Dickey  <dickey@invisible-island.net>

	* package/cm_tools-12.0.spec, package/debian/changelog: bump

	* user/rcsput.man, user/rcsget.man, user/permit.man, user/link2rcs.man,
	  user/checkup.man, user/checkin.man, user/copy.man, user/checkout.man,
	  user/baseline.man:
	check-manpage fixes (rename Ex to Eh for consistency, other errata)

2019-12-05  Thomas E. Dickey  <dickey@invisible-island.net>

	* package/cm_tools-12.0.spec, package/debian/changelog: bump

	* src/checkin/src/rcsput.c, src/checkin/src/checkin.c:
	use DYN-argv lists to replace catarg()

	* src/vcs/src/vcs.c, src/permit/src/permit.c:
	use show_argv2(), to handle verb vs argv[0]


src/checkout/src/checkout.c
/*
 * Title:	checkout.c (front end for RCS checkout)
 * Author:	T.E.Dickey
 * Created:	20 May 1988 (from 'sccsdate.c')
 * Modified:
 *		05 Dec 2019, use DYN-argv lists.
 *		14 Dec 2014, coverity warnings
 *		04 Sep 2012, correction to previous -c change.
 *		23 Oct 2005, correct parsing if -c param is not a separate
 *			     option.
 *		28 Dec 2000, restore file-ownership if setuid'd to root.
 *		12 Nov 1994, pass-thru '-f', '-k' options to 'ci'.
 *		22 Sep 1993, gcc warnings
 *		24 Jun 1993, fixes for apollo-setuid for RCS version 5.
 *		02 Nov 1992, mods for RCS version 5.
 *		16 Jul 1992, corrected call on 'cutoff()'
 *		06 Feb 1992, revise filename-parsing with 'rcsargpair()',
 *			     obsoleted "-x" option.
 *		17 Jan 1992, make this interpret "-p" option.
 *		21 Oct 1991, corrected uid-use in pre-processing.  Handle case
 *			     in which user has write-access in the RCS directory
 *			     even if he owns none of the files.
 *		09 Oct 1991, convert to ANSI. Correct non-root setuid in the
 *			     case where real user has proper permissions.
 *			     (i.e., don't do rcstemp-hack).
 *		08 Oct 1991, stifle message about 'revert()' if root-setuid.
 *		06 Sep 1991, modified interface to 'rcsopen()'
 *		11 Jul 1991, make this work properly with suid-root
 *		20 Jun 1990, use 'shoarg()'
 *		20 May 1991, mods to compile on apollo sr10.3
 *		19 Apr 1990, corrected so we don't pass "-x" to 'co'.
 *			     added "-x" option (to help with makefiles, etc)
 *		10 Oct 1989, corrected last change (if no directory name is
 *			     explicit, use ".")
 *		24 Aug 1989, exit with error if 'usercopy()' fails.  Suppress
 *			     'revert()' message if user does not want to lock
 *			     the file.  Also, verify that working file's
 *			     directory exists (otherwise 'usercopy()' silently
 *			     fails...)
 *		22 Aug 1989, corrected format of debug-message
 *		29 Mar 1989, if we cannot find the archive, this may be because
 *			     the user has hidden his directory (and therefore
 *			     we cannot see it in setuid mode).  Revert to normal
 *			     rights in this case and try again.
 *		28 Mar 1989, forgot to invoke 'rcspermit()' to check directory-
 *			     level permit.
 *		21 Mar 1989, strip setuid privilege if caller wants to check out
 *			     a file not owned by the setuid process.
 *		24 Jan 1989, don't insist that archive be owned by user or euid
 *			     if no lock is being made.  Corrected hole which
 *			     caused working file to be deleted if setuid-use
 *			     prompted for overwrite which was rejected.
 *		14 Dec 1988, use 'vercmp()' rather than 'dotcmp()', to make
 *			     retrieval of increments within a baseline get a
 *			     correct timestamp.
 *		06 Dec 1988, added some DEBUG-traces.
 *		13 Sep 1988, added cleanup handler.  Refined permission-checks.
 *		09 Sep 1988, use 'rcspath()'
 *		30 Aug 1988, broke out 'userprot()'.
 *		25 Aug 1988, check for and accommodate 'setuid()' usage.
 *		24 Aug 1988, added 'usage()' message.  Implemented '-c' cutoff,
 *			     and '-w', '-s' options.  If user has locked RCS
 *			     file, make the checked-out file writeable.
 *		13 Jun 1988, use 'newzone()'.
 *		08 Jun 1988, more adjustments to 'chmod()'.
 *		07 Jun 1988, make this set the checked-out file's mode to
 *			     readonly unless it is locked.  (The 'co' utility
 *			     neglects to do this & is needed on apollo).
 *		27 May 1988, recoded using 'rcsedit' module.
 *
 * Function:	Make a package around the RCS utility 'co', added the ability
 *		to manipulate the checkin date, and file ownership.
 *
 * Options:	This passes through to 'co' all options except the '-d' (date)
 *		option -- we interpret a '-c' in SCCS-style to simplify the
 *		implementation.
 *
 * patch:	This does not interpret branches.
 *
 *		If the user has locked a version other than the tip, and then
 *		does a 'co' without the version, 'co' will still give the
 *		tip-version.  Should provide the revision-option to 'co' myself.
 *		For now, simply assume that locks are on the tip version.
 */

#define		ACC_PTYPES
#define		SIG_PTYPES
#define		STR_PTYPES
#include	<ptypes.h>
#include	<rcsdefs.h>
#include	<dyn_str.h>

#include	<errno.h>
#include	<ctype.h>
#include	<signal.h>
#include	<time.h>

MODULE_ID("$Id: checkout.c,v 11.21 2025/01/07 00:54:05 tom Exp $")

/* local definitions */
#define	TELL	if (!silent) FPRINTF
#define	DEBUG(s)	if(debug) FPRINTF s

#define	EMPTY(s)	(s == NULL || *s == EOS)
#define	MISMATCH(a,b)	((!EMPTY(a)) && strcmp(a,b))

#define	REVSIZ	80

#define	CO_TOOL	"co"

static time_t opt_date;
static int silent;
static int debug;		/* set from environment RCS_DEBUG */
static int no_op;		/* suppress actual "co" invocation */
static int locked;		/* TRUE if user is locking file */
static FILE *log_fp;		/* normally stdout, unless "-p" */
static int to_stdout;		/* TRUE if 'co' writes to stdout */

static uid_t Effect, Caller;	/* effective/real uid's */
static char Working[MAXPATHLEN];	/* current names we are using */
static char Archive[MAXPATHLEN];
static char rev_buffer[REVSIZ];
static const char *UidHack;	/* intermediate file for setuid */
static char *opt_rev;		/* revision to find */
static char *opt_who;		/* "-w[login] value     */
static char *opt_sta;		/* "-s[state] value     */
static char *co_f_opt;		/* "-f" option, if any  */
static char *co_k_opt;		/* "-k" option, if any  */

/************************************************************************
 *	local procedures						*
 ************************************************************************/

static void
noPERM(char *name)
{
    errno = EPERM;
    failed(name);
}

static void
noFILE(char *name)
{
    FPRINTF(stderr, "?? file not found: \"%s\"\n", name);
    exit(FAIL);
}

static void
WhoAmI(void)
{
    if (debug)
	show_uids(stdout);
}

/*
 * Cleanup handler
 */
static void
clean_file(void)
{
    if ((UidHack != NULL && *UidHack)
	&& (*Working != 0)
	&& strcmp(UidHack, Working)) {
	(void) unlink(UidHack);
    }
    UidHack = NULL;
}

static SIG_T
cleanup(int sig)
{
    (void) signal(sig, SIG_IGN);
    FPRINTF(stderr, "checkout: cleaning up\n\n");
    clean_file();
    (void) exit(FAIL);
}

static int
TestAccess(char *name, int flag)
{
    int code = access(name, flag);
    DEBUG((log_fp, ".. access(%s,%s) = %d\n", name, access_mode(flag), code));
    return code;
}

static char *
path_of(char *dst, char *src)
{
    char *s = strrchr(strcpy(dst, src), '/');
    if (s == NULL)
	(void) strcpy(dst, ".");
    else
	*s = EOS;
    return dst;
}

static void
GiveBack(int tell_why, const char *why)
{
    if (revert((tell_why || debug) ? why : (char *) 0)) {
	Effect = geteuid();
	WhoAmI();
    }
}

/*
 * Trim the portion of a version off that 'same_branch()' does not match.
 */
static void
trim_branch(char *s)
{
    char *t = strrchr(s, '.');
    if (t != NULL) {
	*t = EOS;
	if (strrchr(s, '.') != NULL) {
	    ;
	} else
	    *s = EOS;		/* trunk (i.e., "9.1") */
    } else
	*s = EOS;		/* trunk (i.e., "9") */
}

/*
 * Check two versions to see that they are on the same branch.
 */
static int
same_branch(char *a, char *b)
{
    char temp1[BUFSIZ], temp2[BUFSIZ];
    trim_branch(strcpy(temp1, a));
    trim_branch(strcpy(temp2, b));
    return !strcmp(temp1, temp2);
}

/*
 * Find the revision which the user has selected.
 */
static int
PreProcess(time_t *revtime,	/* date with which to touch file */
	   int *co_mode)
{
    int ok_vers = FALSE, ok_date = FALSE;
    char key[BUFSIZ];
    char tmp[BUFSIZ];
    char this_rev[REVSIZ];
    char *s = NULL;
    int header = TRUE, code = S_FAIL;

    if (!rcsopen(Archive, -debug, TRUE)) {
	FPRINTF(stderr, "? cannot open archive %s\n", Archive);
	return FALSE;
    }

    *rev_buffer = EOS;
    if (!EMPTY(opt_rev))
	(void) strcpy(rev_buffer, opt_rev);

    while (header && (s = rcsread(s, code))) {
	s = rcsparse_id(key, s);

	switch (code = rcskeys(key)) {
	    /*
	     * Begin an admin description.  If the user did not specify a
	     * revision, we assume the 'tip' version, unless he had one
	     * locked, in which case we assume that.
	     */
	case S_HEAD:
	    s = rcsparse_num(this_rev, s);
	    if (*rev_buffer == '\0')
		(void) strcpy(rev_buffer, this_rev);
	    break;
	case S_SYMBOLS:
	    s = rcssymbols(s, rev_buffer, rev_buffer);
	    break;
	case S_LOCKS:
	    /* see if this was locked by the user */
	    (void) strcpy(tmp, getuser());
	    *this_rev = EOS;
	    s = rcslocks(s, tmp, this_rev);
	    if (*this_rev && EMPTY(opt_rev)) {
		TELL(log_fp, "** revision %s is locked\n", this_rev);
		*co_mode |= S_IWRITE;
		(void) strcpy(rev_buffer, this_rev);
	    }
	    break;

	    /*
	     * Begin a delta description.  We are looking (like 'co') for
	     * the last version along a branch which matches the cutoff,
	     * state, and who-options.  Since the rcs file stores deltas
	     * in reverse-order, the first one we can match the options
	     * against is the right one.
	     */
	case S_VERS:
	    (void) sprintf(this_rev, "%.*s", REVSIZ - 1, key);
	    DEBUG((log_fp, "version = %s\n", this_rev));
	    ok_vers = same_branch(rev_buffer, this_rev);
	    ok_date = FALSE;
	    break;
	case S_DATE:
	    s = rcsparse_num(tmp, s);
	    if (ok_vers) {
		*revtime = rcs2time(tmp);
		DEBUG((log_fp, "date    = %s", ctime(revtime)));
		ok_date = ((opt_date == 0)
			   || (*revtime <= opt_date));
	    }
	    break;
	case S_AUTHOR:
	    s = rcsparse_id(key, s);
	    if (ok_vers && MISMATCH(opt_who, key))
		ok_vers = FALSE;
	    DEBUG((log_fp, "author  = %s\n", key));
	    break;
	case S_STATE:
	    s = rcsparse_id(key, s);
	    DEBUG((log_fp, "state   = %s\n", key));
	    if (ok_vers && MISMATCH(opt_sta, key))
		ok_vers = FALSE;
	    break;

	    /* 'next' is the last keyword in a delta description */
	case S_NEXT:
	    if (ok_vers && ok_date) {
		if (EMPTY(opt_rev)) {
		    if (strcmp(rev_buffer, this_rev)) {
			*co_mode &= ~S_IWRITE;
		    }
		    (void) strcpy(rev_buffer, this_rev);
		    header = FALSE;
		    break;
		}
		DEBUG((log_fp, "compare %s %s => %d (for equality)\n",
		       this_rev, rev_buffer,
		       vercmp(this_rev, rev_buffer, TRUE)));
		if (vercmp(this_rev, rev_buffer, TRUE) == 0) {
		    (void) strcpy(rev_buffer, this_rev);
		    header = FALSE;	/* force an exit */
		}
	    }
	    break;
	case S_DESC:
	    header = FALSE;
	}
	if (!s)
	    break;
    }
    rcsclose();

    if (!(ok_vers && ok_date)) {
	FPRINTF(stderr, "? cannot match requested revision\n");
	return FALSE;
    }
    return TRUE;
}

/*
 * Do the actual check-out.  For RCS version 5, we must always do this as admin,
 * since the 'ci' program gets confused by the apollo set-uid.
 */
static int
RcsCheckout(void)
{
    ARGV *args;
    const char *opt = to_stdout ? "-p" : (locked ? "-l" : "-r");
    int code = 0;

    args = argv_init2(CO_TOOL, rcspath(CO_TOOL));
#if RCS_VERSION >= 5
    argv_append(&args, "-M");
#endif
    if (silent)
	argv_append(&args, "-q");
    argv_append2(&args, opt, rev_buffer);
    if (!EMPTY(co_f_opt))
	argv_append(&args, co_f_opt);
    if (!EMPTY(co_k_opt))
	argv_append(&args, co_k_opt);
    argv_append(&args, UidHack);
    argv_append(&args, Archive);

    if (!silent || debug)
	show_argv2(log_fp, CO_TOOL, argv_values(args));
    if (!no_op) {
	Stat_t sb;
	if (Effect == 0) {
	    if (stat(Archive, &sb) < 0)
		sb.st_uid = sb.st_gid = 0;
	}
	code = executev(argv_values(args));
	if (Effect == 0
	    && Effect != sb.st_uid)
	    chown(Archive, sb.st_uid, sb.st_gid);
    }
    argv_free(&args);
    return (code);
}

/*
 * Check out the file using the RCS 'co' utility.  If 'co' does something, then
 * it will delete or modify the checked-in file.
 */
static void
Execute(time_t newtime, time_t oldtime, int co_mode)
{
    int code;
    Stat_t sb;
    long oldctime;
    int copied;

    UidHack = to_stdout ? "" : rcstemp(Working, FALSE);
    copied = strcmp(UidHack, Working);
    oldctime = 0;
    if (!copied && stat(UidHack, &sb) >= 0)
	oldctime = sb.st_ctime;

#if	RCS_VERSION >= 5
    if (saves_uid()) {
	code = RcsCheckout();
    } else {
	code = for_admin(RcsCheckout);
    }
#else
    code = RcsCheckout();
#endif

    if ((code >= 0) && !no_op && !to_stdout) {
	if (stat(UidHack, &sb) >= 0) {
	    DEBUG((log_fp, "=> file \"%s\"\n", UidHack));
	    DEBUG((log_fp, "=> size = %ld\n", (long) (sb.st_size)));
	    DEBUG((log_fp, "=> date = %s", ctime(&sb.st_mtime)));
	    DEBUG((log_fp, "..(comp)  %s", ctime(&oldtime)));
	    if (!copied && (sb.st_ctime == oldctime)) {
		TELL(log_fp, "** checkout was not performed\n");
	    } else {
		if (copied && (usercopy(UidHack, Working) < 0))
		    failed(Working);
		else if (userprot(Working, co_mode, newtime) < 0)
		    noPERM(Working);
	    }
	}
    }
    clean_file();
}

/*
 * Process a single file.
 *
 * Before checkout, verify that the file exists, and obtain its modification
 * time/date.  Also (since this looks at the working/archive files before
 * invoking 'co'), see if we have nominal permissions to work upon the files:
 *	a) If the working file exists, we need rights to recreate it with the
 *	   same ownership.  If it does not exist, we assume that we create it
 *	   with the ownership implied from the directory.
 *	b) We need rights to recreate the archive file with its current
 *	   ownership -- this implies that our effective uid must be the same
 *	   as that of the archive.
 *
 * patch: this does not address the need to down-adjust privilege if we start
 *	out with root-uid.
 */
static void
do_file(void)
{
    Stat_t sb;
    int ok;
    char temp[MAXPATHLEN];
    time_t revtime;
    int co_mode;		/* mode with which 'co' sets file */

    /*
     * Ensure that we can find the RCS-file.  Note that this program may be
     * setuid'd to a user who has no rights in the working directory...
     */
    ok = (rcs_archive(Archive, &sb) >= 0);
    if (!ok && (errno != EISDIR)) {
	if (TestAccess(Archive, R_OK) == 0) {
	    GiveBack(FALSE, "directory access");
	    ok = (rcs_archive(Archive, &sb) >= 0);
	}
    }
    if (!ok)
	noFILE(Archive);

    co_mode = (sb.st_mode & 0777);
    if (locked)
	co_mode |= S_IWRITE;
    else
	co_mode &= 0555;	/* strip writeable mode */
    if (!co_mode)
	co_mode = 0400;		/* leave at least readonly! */

    /*
     * If we have the file, ensure that we have proper access:
     */
    if (Effect != Caller) {
	if (locked
	    && (TestAccess(path_of(temp, Archive), W_OK) >= 0))
	    GiveBack(FALSE, "RCS-directory is writable");
	else if (!rcspermit(temp, (char *) 0, (const char **) 0))
	    GiveBack(Effect && locked, "not listed in permit-file");
    }

    /*
     * Check to see if the working-file exists
     */
    if (rcs_working(Working, &sb) >= 0) {
	DEBUG((log_fp, "=> date = %s", ctime(&sb.st_mtime)));
    } else if (errno == EISDIR) {
	noFILE(Working);
    }

    if (!locked && TestAccess(Archive, R_OK) >= 0)
	GiveBack(FALSE, "normal rights suffice");

    if (PreProcess(&revtime, &co_mode))
	Execute(revtime, sb.st_mtime, co_mode);
}

static void
usage(void)
{
    static const char *tbl[] =
    {
	"Usage: checkout [options] [working_or_archive [...]]"
	,""
	,"Options (from \"co\"):"
	,"  -l[rev]  locks the checked-out revision for the caller."
	,"  -p[rev]  prints the revision on standout-output"
	,"  -q[rev]  quiet mode"
	,"  -r[rev]  retrieves the latest revision whose number is less than or equal"
	,"           to \"rev\"."
	,"  -cdate   retrieves the latest revision on the selected branch whose checkin"
	,"           date/time is less than or equal to \"date\", in the format"
	,"                   yy/mm/dd hh:mm:ss"
	,"  -sstate  retrieves the latest revision on the selected branch whose state"
	,"           is set to state."
	,"  -w[login] retrieves the latest revision on the selected branch which was"
	,"           checked in by user \"login\"."
	,""
	,"Unimplemented \"co\" options:"
	,"  -jjoinlist generates a new revision which is the join of the revisions on"
	,"           joinlist"
	,""
	,"Special:"
	,"  -D       debug/no-op"
    };
    unsigned j;
    for (j = 0; j < sizeof(tbl) / sizeof(tbl[0]); j++)
	FPRINTF(stderr, "%s\n", tbl[j]);
    FFLUSH(stderr);
    (void) exit(FAIL);
}

/************************************************************************
 *	main program							*
 ************************************************************************/
/*ARGSUSED*/
_MAIN
{
    int j;
    char *s;

    char who[BUFSIZ], tmp[BUFSIZ];
    int code;

    debug = RCS_DEBUG;
    Caller = getuid();
    Effect = geteuid();
    if (Caller != Effect)
	catchall(cleanup);

    log_fp = stdout;
    WhoAmI();

    /* process options */
    for (j = 1; (j < argc) && (*(s = argv[j]) == '-'); j++) {

	(void) strcpy(tmp, s++);
	code = *s++;

	switch (code) {
	case 'D':
	    no_op++;
	    break;

	case 'c':
	    if (*s != EOS) {
		optind = j;
		argv[j] = s;
		optarg = s;
	    } else {
		optind = j + 1;
		optarg = argv[optind];
	    }
	    opt_date = cutoff(argc, argv);
	    j = optind - 1;
	    FORMAT(tmp, "-d%s", ctime(&opt_date));
	    (void) strtrim(tmp);
	    TELL(log_fp, "++ cutoff: %s", ctime(&opt_date));
	    break;

	case 'f':
	    co_f_opt = argv[j];
	    break;

	case 'k':
	    co_k_opt = argv[j];
	    break;

	case 'q':
	    silent++;
	    if (!EMPTY(s))
		opt_rev = s;
	    break;

	case 'l':
	    locked++;
	    /* FALL-THROUGH */
	case 'r':
	    if (!EMPTY(s))
		opt_rev = s;
	    break;

	case 'p':
	    to_stdout = TRUE;
	    log_fp = stderr;
	    if (!EMPTY(s))
		opt_rev = s;
	    break;

	case 's':
	    if (!EMPTY(s))
		opt_sta = s;
	    else
		usage();
	    DEBUG((log_fp, ">state:%s\n", opt_sta));
	    break;

	case 'w':
	    if (!EMPTY(s))
		opt_who = s;
	    else
		opt_who = strcpy(who, getuser());
	    DEBUG((log_fp, ">who:%s\n", opt_who));
	    break;

	default:
	    FPRINTF(stderr, "?? Unknown option: %s\n", argv[j]);
	    usage();
	}
    }
    DEBUG((log_fp, ">rev:%s\n", opt_rev));

    /* process files */
    if (j < argc) {
	while (j < argc) {
	    j = rcsargpair(j, argc, argv);
	    do_file();
	}
    } else {
	FPRINTF(stderr, "? expected a filename\n");
	usage();
    }

    (void) exit(SUCCESS);
    /*NOTREACHED */
}
